internal/wasi/fs.go

```diff
@@ -5,100 +5,137 @@ import (
+	dir := &memFSEntry{Entries: m}
+
+	// Path "." indicates m itself as a directory.
+	if name == "." {
+		return newMemDir(dir, "."), nil
+	}
+
+	// Traverse directories.
+	files := strings.Split(name, "/")
+	for _, file := range files[:len(files)-1] {
+		var ok bool
+		dir, ok = dir.Entries[file]
+		if !ok || !dir.IsDir() {
+			return nil, &os.PathError{Op: "open", Path: name, Err: fs.ErrNotExist}
 		}
+	}
 
-		if oFlags&wasi.O_TRUNC == 0 {
-			buf = append(buf, bts...)
+	// Open or create the entry.
+	// Note: MemFS does not check permission for now.
+	baseName := path.Base(name)
+	entry, ok := dir.Entries[baseName]
+	if !ok {
+		if flag&os.O_CREATE == 0 {
+			return nil, &os.PathError{Op: "open", Path: name, Err: fs.ErrNotExist}
 		}
+
+		// Create a new entry.
+		entry = &memFSEntry{}
+		dir.Entries[baseName] = entry
+	} else if flag&os.O_EXCL != 0 && flag&os.O_CREATE != 0 {
+		// when O_EXCL is set with O_CREATE, the file must not already exist.
+		return nil, &os.PathError{Op: "open", Path: name, Err: fs.ErrExist}
 	}
 
-	ret := &memFile{buf: buf}
+	if flag&os.O_TRUNC != 0 {
+		entry.Contents = []byte{}
+	}
 
-	if fsRights&wasi.R_FD_WRITE != 0 {
-		ret.flush = func(bts []byte) {
-			m.Files[path] = bts
-		}
+	if entry.IsDir() {
+		return newMemDir(entry, baseName), nil
+	} else {
+		return &memFile{name: baseName, fsEntry: entry}, nil
 	}
+}
+
+// Open implements fs.FS.Open.
+func (m *MemFS) Open(path string) (fs.File, error) {
+	return m.OpenFile(path, os.O_RDONLY, 0)
+}
 
-	return ret, nil
+// memFSEntry represents a file or directory entry of memFS.
+type memFSEntry struct {
+	// Contents of the file if this file is a regular file.
+	Contents []byte
+	// Entries if this file is a directory. Otherwise, nil.
+	Entries map[string]*memFSEntry
 }
 
-// memFile implements wasi.File
+// IsDir returns if this entry is a directory.
+func (e *memFSEntry) IsDir() bool { return e.Entries != nil }
+
+// memFile represents an opened memFS regular file.
+// memFile implements fs.File, io.Writer, and io.Seeker.
 type memFile struct {
-	buf    []byte
+	// name of the opened file, not a full path but a base name.
+	name string
+	// Seek offset of this opened file.
 	offset int64
-	flush  func(bts []byte)
+	// The memFSEntry this memFile is opened for.
+	fsEntry *memFSEntry
 }
 
-// Read implements io.Reader
+// Stat implements fs.File.Stat
+func (f *memFile) Stat() (fs.FileInfo, error) {
+	return &memFileInfo{name: f.name, fsEntry: f.fsEntry}, nil
+}
+
+// Read implements fs.File.Read
 func (f *memFile) Read(p []byte) (int, error) {
 	// In memFile, the end of the buffer is the end of the file.
-	if f.offset == int64(len(f.buf)) {
+	if f.offset == int64(len(f.fsEntry.Contents)) {
 		return 0, io.EOF
 	}
-	nread := copy(p, f.buf[f.offset:])
+	nread := copy(p, f.fsEntry.Contents[f.offset:])
 	f.offset += int64(nread)
 	return nread, nil
 }
 
+// Close implements fs.File.Close
+func (f *memFile) Close() error {
+	return nil
```

Comment:
trailing comment on no-op, but maybe we want to nil its body to allow faster GC on close?